### PR TITLE
Add Mul postprocessor for Vector

### PR DESCRIPTION
**Pull request by jhaayushkumar**
#### References to other Issues or PRs

Related to gh-26121 and gh-26623.

#### Brief description of what is fixed or changed

When [simplify()](cci:1://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/tests/test_vector.py:193:0-225:60) traverses expression trees using `bottom_up()`, it reconstructs nodes via `.func(*args)`. This caused [VectorMul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) objects to become [Mul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) objects, which then failed type checks in `CoordSys3D.__new__()` with `TypeError: location should be a Vector`.

This PR adds a postprocessor for [Mul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) (similar to the existing [Add](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:498:0-517:27) postprocessor from gh-26623) to ensure [Mul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) operations on vectors return [VectorMul](cci:2://file:///Users/ayushkumarjha/Desktop/sympy/sympy/vector/vector.py:520:0-539:35) objects.

#### Release Notes

* vector
  * Fixed TypeError when simplifying vector expressions. VectorMul objects are now preserved during simplification.

### CODE REVIEW

**Comment by jhaayushkumar** [comment]
I'm not sure about this one but the understanding that i gained from  failing of `object has no attribute '_components'. Did you mean: 'comp...`
```python
=============================== short test summary info ===============================

FAILED sympy/printing/tests/test_latex.py::test_latex_vector_expressions - AssertionError: assert '\\mathbf{\\hat{k}_{A}}' == 'x \\left(\\m..._{A}}\\right)'

FAILED sympy/printing/tests/test_mathml.py::test_print_Vector - assert '<msub><mover...A</mi></msub>' == '<mrow><mi>x<.../mrow></mrow>'

FAILED sympy/printing/pretty/tests/test_pretty.py::test_vector_expr_pretty_printing - AssertionError: assert 'k_A' == 'x⋅(i_A)×(j_A)'

FAILED sympy/vector/tests/test_field_functions.py::test_mixed_coordinates - assert Divergence(Cross(a.i, b.j)) == ((((((a.x * b.x) * c.x) * Divergence(Cross(a...

FAILED sympy/vector/tests/test_vector.py::test_cross - AttributeError: 'Cross' object has no attribute '_components'. Did you mean: 'comp...

= 5 failed, 12901 passed, 682 skipped, 329 xfailed, 2 xpassed, 8 warnings in 261.22s (0:04:21) =
```
`Cross(Vector)` made code expect `a _components` attribute that didn't exist, causing crashes.
so i changed `Cross(Vector)` to `Cross(Expr)` - now it's just an unevaluated formula (like sin(x)) not a computed vector, so no `_components` needed and everything works.

**Comment by oscarbenjamin** [comment]
> now it's just an unevaluated formula (like sin(x)) not a computed vector, so no `_components` needed and everything works.

I don't think everything works. The Cross class is broken because it can't work properly with the way that Vector generally works.

The way that Vector generally works is that all vectors are expressed as a linear combination of scalar expressions multiplied by base vector expressions. The unevaluated Cross breaks this by not having that representation (no `_components`).

With this PR we have
```python
In [1]: from sympy.vector import *

In [2]: C = CoordSys3D('C')

In [3]: v = Cross(2*C.i, C.i + C.j)

In [4]: v
Out[4]: ((2) i_C)×(i_C + j_C)

In [5]: 1 + v
Out[5]: 1 + ((2) i_C)×(i_C + j_C)
```
However `1 + v` should not be allowed because that is adding a scalar and a vector.

This is the proper behaviour
```python
In [6]: v.doit()
Out[6]: (2) k_C

In [7]: 1 + v.doit()
---------------------------------------------------------------------------
TypeError
```
I think that the proper fix for `Cross` is just to have it always evaluate. That defeats the purpose of having the `Cross` class at all but I think adding that class was misguided since it does not work properly in general.

File: sympy/vector/vector.py [hunk 1]
```diff
@@ -418,19 +418,22 @@ def _div_helper(one, other):
...
     def _postprocessor(expr):
-        vec_class = {Add: VectorAdd}[cls]
+        vec_class = {Add: VectorAdd, Mul: VectorMul}[cls]
         vectors = []
         for term in expr.args:
             if isinstance(term.kind, VectorKind):
                 vectors.append(term)
 
         if vec_class == VectorAdd:
             return VectorAdd(*vectors).doit(deep=False)
```
Comment (Davide-sd) [inline sympy/vector/vector.py]: I don't think this is correct. @oscarbenjamin look at this:

```
from sympy import Add
from sympy.vector import CoordSys3D
C = CoordSys3D('C')
Add(2, C.i)
# out: C.i
```

In my opinion, all arguments should go to VectorAdd, which should then validate them and decide what to do.

**Comment by Davide-sd** [comment]
Also note that other classes like `Curl, Gradient, Laplacian` inherits from `Expr` and suffer from the same behavior shown above by @oscarbenjamin .

**Comment by Davide-sd** [comment]
@Upabjojr I'm going trough the commit history of the vector module in order to understand a few design choices. I noticed that you worked on it, both as a developer and as a mentor to a GSoC. I have a couple of questions:

* Classes like `Curl, Gradient` were introduced in #12696 . Originally, they inherited from `Basic`, but then they changed to `Expr`. Do you remember the rationale about that choice?
* The `Cross` class was meant to represent an unevaluated cross product. Then, at some point you added [this "evaluation"](https://github.com/Davide-sd/sympy/commit/12933f6704ede1b4ede45a1e8418e900ad1f65ef#diff-02bde2b36adf1a3cdd3c3b9dc5b634710deeb17940f6e03f117fa1db897e8136R437). When I'm creating an unevaluated object, I would expect to get that unevaluated object. But here, under a certain condition, I get a different object... Do you remember the rationale about that choice?

**Comment by oscarbenjamin** [comment]
> other classes like `Curl, Gradient, Laplacian` inherits from `Expr` and suffer from the same behavior

While that is not ideal it probably means that having Cross do the same is a good solution for now in terms of this PR.

**Comment by Upabjojr** [comment]
> @Upabjojr I'm going trough the commit history of the vector module in order to understand a few design choices. I noticed that you worked on it, both as a developer and as a mentor to a GSoC. I have a couple of questions:
> 
> * Classes like `Curl, Gradient` were introduced in [Create operators.py file #12696](https://github.com/sympy/sympy/pull/12696) . Originally, they inherited from `Basic`, but then they changed to `Expr`. Do you remember the rationale about that choice?
> * The `Cross` class was meant to represent an unevaluated cross product. Then, at some point you added [this "evaluation"](https://github.com/Davide-sd/sympy/commit/12933f6704ede1b4ede45a1e8418e900ad1f65ef#diff-02bde2b36adf1a3cdd3c3b9dc5b634710deeb17940f6e03f117fa1db897e8136R437). When I'm creating an unevaluated object, I would expect to get that unevaluated object. But here, under a certain condition, I get a different object... Do you remember the rationale about that choice?

Don't remember right now. You usually inherit from `Expr` if the expression is a mathematical formula, which is the case here, apparently.

Concerning the constructor, most of SymPy suffers from this kind of problem. Even `Add` and `Mul` do some automatic evaluation, unfortunately. The point is that forcing the expression tree into a standard canonical form make developing the code easier, as you don't have to take into account non-standard forms.

**Comment by jhaayushkumar** [comment]
> While that is not ideal it probably means that having Cross do the same is a good solution for now in terms of this PR.


@oscarbenjamin  sorry for tag but sir, I need help, should i fix the cross first as, Cross does not work in a VectorMul is a bug but it should be fixed by changing Cross somehow. or try another approach
as the fix suggested here https://github.com/sympy/sympy/issues/28727#issuecomment-3634196062  failed 6 test cases 
Screenshot 2025-12-22 at 1 56 39 AM

**Comment by oscarbenjamin** [comment]
> should i fix the cross first as, Cross does not work in a VectorMul is a bug but it should be fixed by changing Cross somehow.

In this PR you have made Cross a subclass of Expr which solves the failed tests that you showed. A better fix should be found in future for Cross and also Curl etc but I think this approach is fine for now in this PR.

**Comment by jhaayushkumar** [comment]
>but I think this approach is fine for now in this PR.
>
should i squash the commit for merge.

**Comment by oscarbenjamin** [comment]
The comment by @Davide-sd above should be addressed https://github.com/sympy/sympy/pull/28749#discussion_r2637048161

**Comment by oscarbenjamin** [comment]
Also it is not necessary to squash the commits manually any more because the person merging the PR can do it (https://github.com/sympy/sympy/issues/18548#issuecomment-3651000245).

**Comment by jhaayushkumar** [comment]
> The comment by @Davide-sd above should be addressed [#28749 (comment)](https://github.com/sympy/sympy/pull/28749#discussion_r2637048161)


```diff
if vec_class == VectorAdd:
+   if not vectors:
+       return expr
+   if len(vectors) < len(expr.args):
+       return expr
    return VectorAdd(*vectors).doit(deep=False)
```
on addressing the comment i had added two checks first one check whether they are vector or not `if not vectors:` and the second check for a mix of vector and scalars `if len(vectors) < len(expr.args)` did it looks good

**Comment by oscarbenjamin** [comment]
It should just pass all the args to `VectorAdd`. The `VectorAdd` function can decide what to do with them and should raise an error if any are not vectors.

**Comment by jhaayushkumar** [comment]
> It should just pass all the args to `VectorAdd`. The `VectorAdd` function can decide what to do with them and should raise an error if any are not vectors.
>

```diff
if vec_class == VectorAdd:
-    return VectorAdd(*vectors).doit(deep=False)
+    return VectorAdd(*expr.args).doit(deep=False)
def __new__(cls, *args, **options):
+   for arg in args:
+       if not isinstance(arg,Vector):
+           raise TypeError("VectorAdd expects all arguments to be Vector instances")
```

did it ok now?
and sorry for the delay sir, as i was busy in my semester exams.

File: sympy/vector/vector.py [hunk 2]
```diff
@@ -418,19 +418,22 @@ def _div_helper(one, other):
...
         vectors = []
         for term in expr.args:
             if isinstance(term.kind, VectorKind):
                 vectors.append(term)
 
         if vec_class == VectorAdd:
-            return VectorAdd(*vectors).doit(deep=False)
+            return VectorAdd(*expr.args).doit(deep=False)
+        if vec_class == VectorMul:
+            return VectorMul(*vectors).doit(deep=False)
```
Comment (oscarbenjamin) [inline sympy/vector/vector.py]: Can this not just pass all of `expr.args`? What happens to the other args?

I think there needs to be some tests in the `test_vector.py` for what happens when calling `Add(*args)` or `Mul(*args)` with args that are vectors or scalars. These can go along with the tests for kind.

File: sympy/vector/tests/test_vector.py [hunk 3]
```diff
@@ -341,3 +341,73 @@ def test_scalar():
...
+    assert result_rational == Rational(3, 2)*C.k
+    assert result_rational.measure_number == Rational(3, 2)
+
+
+def test_postprocessor_mul_vector_vector():
+    raises(ValueError, lambda: Mul(C.i, C.j))
+    raises(ValueError, lambda: Mul(C.j, C.k))
+
+    result_same = Mul(C.i, C.i)
+    assert result_same == C.i**2
```
Comment (Davide-sd) [inline sympy/vector/tests/test_vector.py]: What is going on here?

Comment (jhaayushkumar) [reply sympy/vector/tests/test_vector.py]: as sympy automatically converts Mul(C.i, C.i) into C.i**2 before our vector code processes it, so the validation logic never reaches it. I added this test to ensure this default behavior stays consistent or any effect occours on it, does it fine or need any refining.

File: sympy/vector/tests/test_vector.py [hunk 4]
```diff
@@ -341,3 +341,73 @@ def test_scalar():
...
+    raises(TypeError, lambda: Add(Rational(3, 2), C.k))
+
+
+def test_postprocessor_mul_scalar_vector():
+    x, y = symbols('x y')
+
+    result = Mul(2, C.i)
+    assert result == 2*C.i
+    assert isinstance(result, VectorMul)
+    assert result.measure_number == 2
```
Comment (jhaayushkumar) [inline sympy/vector/tests/test_vector.py:361]: i added 
```python
assert result.measure_number == 2
```
to insure that scalars are `NOT` being filtered out does it seems good for the filteration test.

File: sympy/vector/tests/test_vector.py (hunk 3 above)
Comment (oscarbenjamin) [reply sympy/vector/tests/test_vector.py]: I don't see why this was resolved. This behaviour does not seem correct to me.

Comment (jhaayushkumar) [reply sympy/vector/tests/test_vector.py]: sympy converts `Mul(C.i, C.i) to C.i**2` before our code runs. I thought since this edge case exists and skip our validation, I should add a test to track it. 
but I removed it because I realized C.i**2 doesn't make mathematical sense for vectors.

Comment (Upabjojr) [reply sympy/vector/tests/test_vector.py]: Mul(C.i, C.i) should probably raise an exception. An alternative would be to follow the overloading of * for two vectors, is there such an operation?

Comment (Davide-sd) [reply sympy/vector/tests/test_vector.py]: Maybe we can add another postprocessor for the `Pow` class. Something like this:

```py
        vec_class = {Add: VectorAdd, Mul: VectorMul, 
            # None because there is no VectorPow class
            Pow: None}
        ...
        elif vec_class is None:
            if isinstance(expr.args[0], Vector):
                raise ValueError("Power operation is not supported for vectors")

...

Basic._constructor_postprocessor_mapping[Vector] = {
    "Add": [get_postprocessor(Add)],
    "Mul": [get_postprocessor(Mul)],
    "Pow": [get_postprocessor(Pow)],
}
```

Comment (jhaayushkumar) [reply sympy/vector/tests/test_vector.py]: Agreed! I was thinking the same. since vectors don't support power operations, so the postprocessor approach makes perfect sense.

Comment (Upabjojr) [reply sympy/vector/tests/test_vector.py]: it's not just power... also `Mul(C.i, C.j)` should not be allowed. Vectors have either dot-product or cross-product (at least in 3D mechanics), they are both operations unrelated to standard multiplication.

Comment (jhaayushkumar) [reply sympy/vector/tests/test_vector.py]: > it's not just power... also `Mul(C.i, C.j)` should not be allowed. Vectors have either dot-product or cross-product (at least in 3D mechanics), they are both operations unrelated to standard multiplication.

i already handled this by:-
```python
def test_postprocessor_mul_vector_vector():
    raises(ValueError, lambda: Mul(C.i, C.j))
    raises(ValueError, lambda: Mul(C.j, C.k))
```

File: sympy/vector/tests/test_coordsysrect.py [hunk 5]
```diff
@@ -497,3 +497,13 @@ def test_cartesian_rotation_transformation_equations():
         xb * cos(alpha) - yb * sin(alpha),
         xb * sin(alpha) + yb * cos(alpha),
         zb)
+
+def test_issue_28727():
```
Comment (oscarbenjamin) [inline sympy/vector/tests/test_coordsysrect.py:764]: There should be two blank lines between top-level functions as in the rest of the file.

**Comment by oscarbenjamin** [comment]
The outstanding issue here is:
```
In [1]: from sympy.vector import CoordSys3D, express

In [2]: C = CoordSys3D('C')

In [3]: Mul(C.i, C.i)
Out[3]: 
   2
i_C 
```

File: sympy/vector/tests/test_vector.py (hunk 3 above)
Comment (Davide-sd) [reply sympy/vector/tests/test_vector.py]: Remember to add a test for `C.i**2`.

**Comment by jhaayushkumar** [comment]
Fixed! the Pow postprocessor catches Mul(C.i, C.i) and raises ValueError. also add test for C.i**2 .

**Comment by oscarbenjamin** [comment]
> catches Mul(C.i, C.i) and raises ValueError.

It should be TypeError.

**Comment by jhaayushkumar** [comment]
> > catches Mul(C.i, C.i) and raises ValueError.
> 
> It should be TypeError.

done!

File: sympy/vector/vector.py [hunk 6]
```diff
@@ -418,19 +418,22 @@ def _div_helper(one, other):
...
+            if isinstance(expr.args[0], Vector):
+                raise TypeError("Power operation is not supported for vectors")
     return _postprocessor
 
 
 Basic._constructor_postprocessor_mapping[Vector] = {
     "Add": [get_postprocessor(Add)],
+    "Mul": [get_postprocessor(Mul)],
+    "Pow": [get_postprocessor(Pow)],
 }
```
Comment (oscarbenjamin) [inline sympy/vector/vector.py]: The `get_postprocessor` function seems like a pointless complication. Better just to have
```
def _add_vector_processor(expr: Add) -> Vector:
    return VectorAdd(*expr.args).doit(deep=False)
```

File: sympy/vector/vector.py [hunk 7]
```diff
@@ -416,21 +416,22 @@ def _div_helper(one, other):
...
-        if vec_class == VectorAdd:
-            return VectorAdd(*vectors).doit(deep=False)
-    return _postprocessor
+def _mul_vector_postprocessor(expr: Mul) -> Vector:
+    return VectorMul(*expr.args).doit(deep=False)
+
+def _pow_vector_postprocessor(expr: Pow):
+    if isinstance(expr.args[0], Vector):
+        raise TypeError("Power operation is not supported for vectors")
+    return expr
```
Comment (oscarbenjamin) [inline sympy/vector/vector.py]: It should also be an error if the second argument is a vector. I think this should just raise TypeError without any checking.

Comment (oscarbenjamin) [reply sympy/vector/vector.py]: A test for this should also be added.

Comment (jhaayushkumar) [reply sympy/vector/vector.py]: added:

```python
def _add_vector_postprocessor(expr: Add) -> Vector:
    return VectorAdd(*expr.args).doit(deep=False)
    
def _mul_vector_postprocessor(expr: Mul) -> Vector:
    return VectorMul(*expr.args).doit(deep=False)

def _pow_vector_postprocessor(expr: Pow):
    raise TypeError("Power operation is not supported for vectors")

```

**Comment by oscarbenjamin** [comment]
I think this looks good now. @Upabjojr or @Davide-sd do you want to review?

File: sympy/vector/vector.py [hunk 8]
```diff
@@ -499,6 +498,9 @@ class VectorAdd(BasisDependentAdd, Vector):
     """
 
     def __new__(cls, *args, **options):
+        for arg in args:
+            if not isinstance(arg, Vector):
+                raise TypeError("VectorAdd expects all arguments to be Vector instances")
```
Comment (Davide-sd) [inline sympy/vector/vector.py]: Not sure if this is really necessary because `BasisDependentAdd.__new__` is already doing these checks.

Comment (jhaayushkumar) [reply sympy/vector/vector.py]: I thought an explicit check would be safer and provide a much clearer error message than the default one.
But agreed, it's redundant, so should i remove it ?

Comment (jhaayushkumar) [reply sympy/vector/vector.py]: remove the redundant part.

Comment (jhaayushkumar) [reply sympy/vector/vector.py]: @Davide-sd did it ok now?

**Comment by Davide-sd** [comment]
Looks good to me.

**Comment by oscarbenjamin** [comment]
Okay thanks all. Let''s get it in.
